### ISSUE TITLE
training in parallel with several threads

### ISSUE THREAD

**Issue by jerabaul29** [reinforceio/tensorforce#480]
Is there a way to use several threads to generate training data?

What I mean is, when an epoch is running, it is generating data that is put in the replay buffer for training. Is it possible to feed the replay buffer from several threads? I.e., for example, to put 5 copies of the same robotic arm and use them all simultaneously to train?

**Comment by AlexKuhnle** [comment]
This is currently not possible. If you're keen to investigate this, I could probably help with some higher-level guidance on how I would go about implementing it in TensorForce, but won't have the capacity to look into it myself anytime soon, unfortunately.

**Comment by jerabaul29** [comment]
@AlexKuhnle Could you provide a few guidance elements so that we can decide if we invest time in this? Where in the code / which kind of coding is required?

@UlysseREGLADE @16Aghnar can you follow this and this if you can help?

**Comment by jerabaul29** [comment]
@mirok maybe something of interest for you also?

**Comment by jerabaul29** [comment]
Would it be possible for example to pass a list of environments to the runner, and program the runner to fill the replay memory buffer using all the environments, and update the networks each time one of these has terminated?

**Comment by AlexKuhnle** [comment]
I would start with the current "distributed_mode" setup, which will probably need the following modifications:

- Local workers should not have their own [memory variables](https://github.com/reinforceio/tensorforce/blob/master/tensorforce/core/memories/queue.py#L71), but use the global model's memory instead
- Global worker should not just act as a [parameter server](https://github.com/reinforceio/tensorforce/blob/master/tensorforce/models/model.py#L408), but perform updates once [the criteria](https://github.com/reinforceio/tensorforce/blob/master/tensorforce/models/memory_model.py#L566) are met (while the local workers should never themselves perform updates).

How does that sound? I think of it as an "extended distributed_mode", and I'm happy to be involved in its development...

**Comment by AlexKuhnle** [comment]
Yes, I can confirm that the memory is loaded as well. (y)

Yes, if you check the distributed_mode setup, each model has an internal "reference" to the global model, and the local model essentially shouldn't have a memory itself, but use the memory of the global model. Unfortunately, this feature is quite "involved" and may require some reading about and experimenting with Distributed TensorFlow (at least if you have not much experience with it, like me).

**Comment by AlexKuhnle** [comment]
Feel free to get in touch on Gitter for more technical discussions.

**Comment by jerabaul29** [comment]
Ok, this is more difficult than I expected.

Otherwise, should we just have a parallel environment that connects to several sockets / simulations? This should not be a problem as long as no lstms are involved to predict using the same network on several simulations simultaneously, isn't t it?

**Comment by AlexKuhnle** [comment]
Unfortunately this is not the case, as both Q-models with next_state and particularly PG-models base their reward calculation on information from other timestep(s). For a quick-and-dirty solution, you could potentially work around it by using `agent.import_experience(...)` [here](https://github.com/reinforceio/tensorforce/blob/master/tensorforce/agents/learning_agent.py#L144) and collect experience in the right format outside of TensorForce (add an episode once it's finished). Similarly, if you look at `agent.observe(...)` [here](https://github.com/reinforceio/tensorforce/blob/master/tensorforce/agents/agent.py#L164), there might be ways to implement the feature in a similar way within TensorForce on a superficial (i.e. Agent-) level, e.g. for a known fixed number of environment instances. The latter would actually be a generally useful feature for TensorForce. Again, feel free to switch to Gitter for more technical discussion on this.

**Comment by jerabaul29** [comment]
Of course you are right. I had forgotten about this.

**Comment by UlysseREGLADE** [comment]
OK, so indeed I've never used distributed graph with Tensorflow, so I'll first read documentation and play around with them, then I'll come back to you.

**Comment by AlexKuhnle** [comment]
I think the threaded runner may also be helpful here, yes. However, I've myself never really looked into this feature (which was contributed at some point), so can't say much here, but let us know if that helps.

**Comment by jerabaul29** [comment]
@16Aghnar how does this look now? Did you get it to work? Can you explain a bit how?

**Comment by AlexKuhnle** [comment]
I think #508 may solve your problem as well, what do you think?
